File: OLD SNACKS/JAVASCRIPT/menstrualCycleApp.js
const prompt = require("Prompt-sync")();
function calculateNextPeriodDate(lastPeriodDate, cycleLength){
	return new Date(lastPeriodDate.getTime() + cycleLength * 24 * 60 * 60 * 1000);
}
function calculateOvulationDate(lastPeriodDate) {
	return new Date(lastPeriodDate.getTime() + 14 * 24 * 60 * 60 * 1000);
}
function isCycleLengthValid(cycleLength) {
	return cycleLength >= 21 && cycleLength <= 35;
}
function isPeriodLengthValid(periodLength) {
	return periodLength >= 3 && periodLength <= 7;
}
function isValidDate(dateString){
	const regex = /^\d{4}-\d{2}-\d{2}$/;
	if (!regex.test(dataString)) return false;
	
	const date = new Date(dateString);
	return date instanceof Date && !NaN(date);
}
function getValidDate() {
	while (true) {
		let lastPeriodInput = prompt("Enter the date of the last period (YYYY-MM-DD):   ");
		try{
			let lastPeriodDate = new Date(lastPeriodInput);

			if(!isNaN(lastPeriodDate.getTime())) {
				return lastPeriodDate;
			} else {
				console.log("Error: Invalid date format. Please enter the format. YYYY-MM-DD");
			}
		}catch {
			console.log("Error: Invalid date format.");
		}
	}
}
function main() {
	let name = prompt("Enter your Name:   ");
	let cycleLength;
	while (true) {
		cycleLength = parseInt(prompt("Enter average cycle Length (days):   "));
		if (isCycleLengthValid(cycleLength)) {
			break;
		} else {
			console.log("Error: cycle length must be between 21 and 35 days. if this is not the case please see the doctor for treatment or Try again:   ")
		}
	}
	
	let periodLength;
	while (true) {
		let periodLength = parseInt(prompt("Enter period length (days):   "));
		if(isPeriodLengthValid(periodLength)) {
			break;
		}else {
			console.log("Error: Period length must be between 3 and 7 days. if this is not the case please see the doctor for treatment or Try again:    ")
		}
	}

	let lastPeriodDate = getValidDate();
	let nextPeriodDate = calculateNextPeriodDate(lastPeriodDate, cycleLength);
	let ovulationDate = calculateOvulationDate(lastPeriodDate);
	

	console.log(`Hello, ${name}!`); 
	console.log(`Your next period is expected on:  ${nextPeriodDate.toISOString().split('T')[0]}`);
	console.log(`Your ovulation date is expected on: ${ovulationDate.toISOString().split('T')[0]}`);
}

main();

File: OLD SNACKS/PYTHON/menstrual_cycle_app.py
from datetime import datetime, timedelta

def calculate_next_period_date(last_period_date, cycle_length):
	return last_period_date + timedelta(days=cycle_length)

def calculate_ovulation_date(last_period_date):
	return last_period_date + timedelta(days=14)

def is_cycle_length_valid(cycle_length):
	return 21 <= cycle_length <= 35

def is_period_length_valid(period_length):
	return 3 <= period_length <= 7

def get_valid_date():
	while True:
		last_period_input = input("Enter the date of the last period (YYYY-MM-DD):   ")
		try:
			return datetime.strptime(last_period_input,  "%Y-%m-%d").date()
		except ValueError:
			print("Error: Invalid date format. Please use the YYYY-MM-DD.")

def main():
	name = input("Enter your name:  ")

	while True: 
		cycle_length = int(input("Enter cycle length (days):    "))
		if is_cycle_length_valid(cycle_length):
			break
		else:
			print("Error: cycle length must be between 21 and 35 days. If this is not the case. Please see the doctor for treatment. Please try again:   ")

	while True:
		period_length = int(input("Enter period length (days):   "))
		if is_period_length_valid(period_length):
			break
		else:
			print("Error: Period length must be between 3 and 7 days. If this is not the case. Please see the doctor for treatment. Please try again:     ")

	last_period_date = get_valid_date()
	next_period_date = calculate_next_period_date(last_period_date, cycle_length)
	ovulation_date = calculate_ovulation_date(last_period_date)
	safe_period_start = last_period_date + timedelta(days=period_length + 1)
	safe_period_end = ovulation_date - timedelta(days=1)

	print(f"Hello,  {name}!")
	print(f"Your next period is expected on:  {next_period_date}")
	print(f"Your ovulation date is expected on:  {ovulation_date}")
	print(f"Your safe period is from:  {safe_period_start} to {safe_period_end}")

if __name__ == "__main__":
	main()
